### DISCUSSION TITLE
dpdata for multiple systems of different frame sizes

### DISCUSSION THREAD

**Discussion by kimiaGF** [deepmodeling/dpdata#248] [Q&A] [answered]
I am attempting to load data from multiple systems, however they all have different frames. This seems to be an issue when trying to append the LabeledSystem to the MultiSystem. How can I get around this?

**Comment by njzjz** [comment, marked as answer]
What's your issue?

**Reply by kimiaGF** [reply]
since the systems are of different sizes both with respect to number of atoms and number of frames, the ms.append(<labeled system>) gives a dimension error. I attempted to consolidate frames into one master file for box.raw, coord.raw , etc. however since each system has a varying number of atoms in it, the number of columns in the force.raw, coord.raw, and virial.raw vary and thus cannot be combines into one file to be loaded by numpy.

**Reply by njzjz** [reply]
You should not put them into the same raw file. Can you provide your script and the detailed error?

**Reply by kimiaGF** [reply]
# Python Script

```
from dpdata import LabeledSystem,MultiSystems
from glob import glob
"""
process multi systems
"""
#searches through every composition directory for their OUTCAR files 
fs = glob('/blue/subhash/kimia.gh/kimia_temp/BO/rundir/garun/temp/*/OUTCAR')  

ms=MultiSystems()
for f in fs:
    try:
        ls=LabeledSystem(f)
    except:
        print(f)
    if len(ls)>0:
        ms.append(ls)

ms.to_deepmd_raw('deepmd')
ms.to_deepmd_npy('deepmd')
```

## Error Readout
Errors out after 1 completed iteration of loop 
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Input In [1], in <module>
     14         print(f)
     15     if len(ls)>0:
---> 16         ms.append(ls)
     18 ms.to_deepmd_raw('deepmd')
     19 ms.to_deepmd_npy('deepmd')

File ~/.conda/envs/MLP/lib/python3.10/site-packages/dpdata/system.py:1158, in MultiSystems.append(self, *systems)
   1156 for system in systems:
   1157     if isinstance(system, System):
-> 1158         self.__append(system)
   1159     elif isinstance(system, MultiSystems):
   1160         for sys in system:

File ~/.conda/envs/MLP/lib/python3.10/site-packages/dpdata/system.py:1168, in MultiSystems.__append(self, system)
   1166 if not system.formula:
   1167     return
-> 1168 self.check_atom_names(system)
   1169 formula = system.formula
   1170 if formula in self.systems:

File ~/.conda/envs/MLP/lib/python3.10/site-packages/dpdata/system.py:1196, in MultiSystems.check_atom_names(self, system)
   1193 if len(new_in_self):
   1194     # Previous atom_name not in this system
   1195     system.add_atom_names(new_in_self)
-> 1196 system.sort_atom_names(type_map=self.atom_names)

File ~/.conda/envs/MLP/lib/python3.10/site-packages/dpdata/system.py:357, in System.sort_atom_names(self, type_map)
    352         self.add_atom_names(new_atoms)
    353     # index that will sort an array by type_map
    354     # a[as[a]] == b[as[b]]  as == argsort
    355     # as[as[b]] == as^{-1}[b]
    356     # a[as[a][as[as[b]]]] = b[as[b][as^{-1}[b]]] = b[id]
--> 357     idx = np.argsort(self.data['atom_names'])[np.argsort(np.argsort(type_map))]
    358 else:
    359     # index that will sort an array by alphabetical order
    360     idx = np.argsort(self.data['atom_names'])

IndexError: index 5 is out of bounds for axis 0 with size 5
```

**Reply by kimiaGF** [reply]
I saw in the DeepPot-SE paper that the number of frames and atoms across different systems does not need to be the same, however I am still getting this same dimensionality error when running the code above.

**Reply by njzjz** [reply]
You may provide a minimal example that will cause the error.

**Reply by kimiaGF** [reply]
https://github.com/kimiaGF/Example_issue.git

here is the link for the repo with the files used to run a small portion of the original multi-system parsing script.

**Reply by njzjz** [reply]
@wanghan-iapcm and @AnguseZhang may need to take a look here - the extracted `atom_types` and `atom_names` of the example system are not correct.

**Reply by kimiaGF** [reply]
I’ve updated the repo with POSCARs. Thanks!

**Comment by wanghan-iapcm** [comment]
@kimiaGF Do you mind if we use one of your poscar '6362_OUTCAR' of your example in our unittest?

**Reply by kimiaGF** [reply]
I just updated the repo with the POSCARs for both systems. Thanks!

**Reply by amcadmus** [reply]
Thanks the PR #250  fixing this issue is under review.

**Reply by kimiaGF** [reply]
Is the issue fixed? I'm still having the same problem

**Reply by njzjz** [reply]
You need to manually install the `devel` branch, i.e.
```sh
git checkout devel
python -m pip install .
```
